### PR TITLE
feat(host.installSupplementalPack)

### PR DESCRIPTION
**Pull request by pdonias**
See vatesfr/xo-web#1460

### CODE REVIEW

File: src/api/host.coffee [hunk 1]
```diff
@@ -260,6 +262,51 @@ stats.resolve = {
...
+  return {
+    $sendTo: yield @registerHttpRequest(handleInstallSupplementalPack, { host, size })
+  }
+
+installSupplementalPack.description = 'installs supplemental pack from ISO file'
+
+installSupplementalPack.params = {
+  host: { type: 'string' }
+  size: { type: 'number' }
+}
```
Comment (julien-f) [inline src/api/host.coffee]: Not necessary to supply the size: you can get it in the handler from the `content-length` header.

File: src/api/host.coffee [hunk 2]
```diff
@@ -260,6 +262,51 @@ stats.resolve = {
...
+  if not sr
+    pbds = @getObject(hostId, 'host').$PBDs
+
+    sr = $find(
+      mapToArray(pbds , (pbd) => @getObject(@getObject(pbd, 'PBD').SR))
+      (sr) -> sr && sr.content_type == 'user' && sr.size - sr.physical_usage >= size
+    )
+
+  if not sr
+    throw new Error('no SR available to store installation file')
```
Comment (julien-f) [inline src/api/host.coffee]: Why not handling the SR directly in `Xapi#installSupplementalPack`?

File: src/xapi/index.js [hunk 3]
```diff
@@ -1105,6 +1105,22 @@ export default class Xapi extends XapiBase {
     return loop()
   }
 
+  async installSupplementalPack (stream, { srId, hostId, size, xapi }) {
+    const sr = this.getObject(srId)
+    const host = this.getObject(hostId)
+
+    const vdi = await xapi.createVdi(size, {
+      sr: sr.$ref,
+      name_label: 'Supplemental pack ISO',
```
Comment (julien-f) [inline src/xapi/index.js]: Add `[XO]` prefix.

File: src/api/host.coffee (hunk 2 above)
Comment (pdonias) [inline src/api/host.coffee]: To keep `Xapi#installSupplementalPack` as simple as possible and to pass as few arguments as possible through `registerHttpRequest`.
Is it better to handle it in `Xapi#installSupplementalPack` anyway?

Comment (julien-f) [inline src/api/host.coffee]: Yes, because we could use it without having to duplicate this logic.

File: src/xapi/index.js [hunk 4]
```diff
@@ -1105,18 +1105,29 @@ export default class Xapi extends XapiBase {
     return loop()
   }
 
-  async installSupplementalPack (stream, { srId, hostId, size, xapi }) {
-    const sr = this.getObject(srId)
-    const host = this.getObject(hostId)
+  async installSupplementalPack (stream, { hostId }) {
+    const fileSize = stream.get('content-length')
```
Comment (julien-f) [inline src/xapi/index.js]: Use `stream.length`.

File: src/api/host.coffee [hunk 5]
```diff
@@ -264,45 +264,31 @@ exports.stats = stats;
...
-  if not sr
-    throw new Error('no SR available to store installation file')
-
   # Timeout seems to be broken in Node 4.
   # See https://github.com/nodejs/node/issues/3319
   req.setTimeout(43200000) # 12 hours
 
   try
-    yield xapi.installSupplementalPack(req, { srId: sr.$id, hostId, size, xapi })
+    yield xapi.installSupplementalPack(req, { hostId })
```
Comment (julien-f) [inline src/api/host.coffee]: Add a `length` property on `req` from `req.headers['content-length']`.

Comment (julien-f) [inline src/api/host.coffee]: Handle the case were this header is missing.

File: src/api/host.coffee [hunk 6]
```diff
@@ -260,6 +262,38 @@ stats.resolve = {
...
+installSupplementalPack = $coroutine ({host}) ->
+  return {
+    $sendTo: yield @registerHttpRequest(handleInstallSupplementalPack, { host })
+  }
+
+installSupplementalPack.description = 'installs supplemental pack from ISO file'
+
+installSupplementalPack.params = {
+  host: { type: 'string' }
+}
```
Comment (julien-f) [inline src/api/host.coffee]: Add a resolve to check for the `admin` permission.

File: src/api/host.coffee [hunk 7]
```diff
@@ -260,6 +262,42 @@ stats.resolve = {
...
+    res.end(format.response(0))
+  catch e
+    res.writeHead(500)
+    res.end(format.error(0, new Error(e.message)))
+
+  return
+
+installSupplementalPack = $coroutine ({host}) ->
+  return {
+    $sendTo: yield @registerHttpRequest(handleInstallSupplementalPack, { host })
```
Comment (julien-f) [inline src/api/host.coffee]: Pass the host id instead.

**Pull request merged**